Force users to change password when it's weak

chdinesh1089 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->


**Testing Plan:** <!-- How have you tested? -->


**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

mateuszmandera commented on zerver/views/auth.py:
```diff
@@ -727,6 +731,10 @@ def login_page(
         # https://github.com/django/django/blob/master/django/template/response.py#L19.
         update_login_page_context(request, template_response.context_data)
+    if 'password' in request.POST and request.user.is_authenticated \
+            and not check_password_strength(request.POST['password']):
```
I think we should rather go with the approach of redirecting in `home_real`, so that we just refuse to even load the app and force the user to change. Perhaps we can
1. If password correct, check its strength in EmailAuthBackend.authenticate and set `.needs_to _change_password`  to True on the `UserProfile`.
2. In home_real, check `user_profile.needs_to _change_password` and if True, redirect to the password reset?
3. Clear that flag when the user changes the password.

Perhaps @timabbott can comment if this sounds reasonable.

chdinesh1089 commented on zerver/tests/test_signup.py:
```diff
@@ -572,6 +572,37 @@ def test_login_bad_password_rate_limiter(self) -> None:
+            self.assert_in_success_response(['Weak credentials. Please Update your password to continue using Zulip.'],
+                                            result)
+
+            # Now change password and verify user isn't redirected to reset page.
+            user_profile.set_password(strong_password)
+            user_profile.save()
+            self.logout()
+            result = self.login_with_return(user_profile.delivery_email, strong_password)
+            result = self.client_get(result.url)
+            self.assertNotEqual(result.status_code, 302)
```
I'm not sure if this is right way to check if the user isn't getting redirected to reset page. Could you confirm @mateuszmandera ?

mateuszmandera replied:
Let's assert what the status_code should be, rather than what it isn't. If the server returned `500` here, the test would pass in its current form, but we wouldn't be happy with that status code. Let's also assert what the url is (`/` I think?)

timabbott commented:
So, the UI looks good here, but I'm not happy with the data model.  I think we don't want to be storing this `weak_password` variable; we only want to offer the password change UI to someone who has just authenticated, which isn't achieved this way.

I think it'd be correct to just have the authentication code path serve the redirect to set a password.

(Also, do we need another `urls.py` entry here?  I'd have thought there'd by a better way to render the reset template passing an extra parameter)

chdinesh1089 commented:
Okay, Could you take a look at https://github.com/zulip/zulip/commit/175c3a0e79aa0bfdb73d631bee00e0232dfec228? It doesn't store any variable in the database but it allows the user to skip changing the password. Do we want to go with something like that but not allow the user to login?

(I tried somethings to avoid the new `urls.py` entry. Since `PasswordResetConfirmView` does a [redirect to the same path it was rendered from, removing the token](https://github.com/django/django/blob/stable/2.2.x/django/contrib/auth/views.py#L278#L279), there doesn't seem to be a better way to avoid that.
One thing we could do is have another variable in URL and a view that allows us to pass that extra parameter but that won't be much different than what we're doing now.
Also tried sending a GET parameter to the URL and wrote a view to handle that but the redirect in `PasswordResetConfirmView.dispatch` is  is clearing the GET parameter.)

mateuszmandera commented on zerver/tests/test_auth_backends.py:
```diff
@@ -544,6 +544,7 @@ def _mock_key(self: RateLimitedAuthenticationByUsername) -> str:
 
         def attempt_authentication(username: str, password: str) -> Optional[UserProfile]:
             request = HttpRequest()
+            request.session = self.client.session
```
In this particular place I don't think setting the session to the client session makes much sense? Could we just generate a fresh session here?

mateuszmandera commented on zproject/backends.py:
```diff
@@ -352,6 +352,7 @@ def authenticate(self, request: HttpRequest, *,
         if user_profile is None:
             return None
         if user_profile.check_password(password):
+            request.session['needs_to_change_password'] = not check_password_strength(password)
```
Hmm, since `session` doesn't make much sense in some contexts, such as clients that operate purely on API requests and their session is useless (perhaps we might actually have requests without a session actually occutrrings at some point in the future?), I wonder if it wouldn't be better to do a conditional `if hasattr(request, 'session'`) here. And if there's no session and password is weak, raise an appropriate json error. Perhaps @timabbott will have some thoughts on this, but this would be my approach.  With this, we wouldn't need the weird `request.sessoin = self.client.session` things in tests.

mateuszmandera replied:
There's also one more weird scenario that we need to handle here. I though I talked about it in the topic in #backend, but I checked and apparently no :disappointed: So, if you dive into the `check_password` codepath in Django, there's a somewhat obscure situation that can happen behind the scenes - iirc, it'll be triggered if the password hashing algorithm is changed - in that case, when the user is logging in, Django first checks the password, if it's correct, Django checks the hash algorithm of the password hash in the database, and if it's the old algorithm, django will run `set_password` for this user, to update the hash to the new algorithm. Something like that. Here's a traceback this can generate if the password is also weak:
```
Traceback (most recent call last):
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/core/handlers/base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/core/handlers/base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "./zerver/views/auth.py", line 777, in login_page
    extra_context=extra_context, **kwargs)(request)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/views/generic/base.py", line 71, in view
    return self.dispatch(request, *args, **kwargs)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/utils/decorators.py", line 45, in _wrapper
    return bound_method(*args, **kwargs)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/views/decorators/debug.py", line 76, in sensitive_post_parameters_wrapper
    return view(request, *args, **kwargs)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/utils/decorators.py", line 45, in _wrapper
    return bound_method(*args, **kwargs)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/utils/decorators.py", line 142, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/utils/decorators.py", line 45, in _wrapper
    return bound_method(*args, **kwargs)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/views/decorators/cache.py", line 44, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/contrib/auth/views.py", line 61, in dispatch
    return super().dispatch(request, *args, **kwargs)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/views/generic/base.py", line 97, in dispatch
    return handler(request, *args, **kwargs)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/views/generic/edit.py", line 141, in post
    if form.is_valid():
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/forms/forms.py", line 185, in is_valid
    return self.is_bound and not self.errors
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/forms/forms.py", line 180, in errors
    self.full_clean()
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/forms/forms.py", line 382, in full_clean
    self._clean_form()
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/forms/forms.py", line 409, in _clean_form
    cleaned_data = self.clean()
  File "./zerver/forms.py", line 341, in clean
    realm=realm, return_data=return_data)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/contrib/auth/__init__.py", line 73, in authenticate
    user = backend.authenticate(request, **credentials)
  File "</srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/decorator.py:decorator-gen-2>", line 2, in authenticate
  File "./zproject/backends.py", line 222, in rate_limit_auth
    result = auth_func(*args, **kwargs)
  File "./zproject/backends.py", line 308, in authenticate
    if user_profile.check_password(password):
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/contrib/auth/base_user.py", line 111, in check_password
    return check_password(raw_password, self.password, setter)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/contrib/auth/hashers.py", line 61, in check_password
    setter(password)
  File "/srv/zulip-venv-cache/4eb7d45882f447b847282956553e9f58eeaeb69e/zulip-py3-venv/lib/python3.5/site-packages/django/contrib/auth/base_user.py", line 107, in setter
    self.set_password(raw_password)
  File "./zerver/models.py", line 1210, in set_password
    raise PasswordTooWeakError
zerver.models.PasswordTooWeakError
```

Long story short, `user_profile.check_password(password)` can raise `PasswordTooWeakError`. We need to handle that gracefully here somehow - since this is such a weird and rare scenario, perhaps an appropriate json error would be good enough? Not sure.

chdinesh1089 replied:
Added a new commit for this case.

mateuszmandera commented on zerver/tests/test_signup.py:
```diff
@@ -571,6 +571,34 @@ def test_login_bad_password_rate_limiter(self) -> None:
+
+            # Now change password and verify user isn't redirected to reset page.
+            user_profile.set_password(strong_password)
+            user_profile.save()
```
Can we just submit data like the form would rather than manually calling set_password?

chdinesh1089 replied:
Okay, changed it. (I did it this way thinking we're not testing the reset password page.)

mateuszmandera commented on zerver/views/home.py:
```diff
@@ -187,6 +188,10 @@ def home_real(request: HttpRequest) -> HttpResponse:
         return accounts_accept_terms(request)
 
+    if request.session.get('needs_to_change_password', False):
+        return HttpResponseRedirect(generate_password_reset_url(user_profile,
```
I think a comment explaining this is important. This block will be very confusing for someone reading `home_realm` code without knowing the story.

chdinesh1089 replied:
Added.

timabbott replied:
I applied this change to clean up the interface; we should never use `.replace()` on URLs as it can lead to security bugs:
```
diff --git a/zerver/forms.py b/zerver/forms.py
index 77b8925772..df5a6764bb 100644
--- a/zerver/forms.py
+++ b/zerver/forms.py
@@ -225,10 +225,11 @@ class LoggingSetPasswordForm(SetPasswordForm):
         return self.user
 
 def generate_password_reset_url(user_profile: UserProfile,
-                                token_generator: PasswordResetTokenGenerator) -> str:
+                                token_generator: PasswordResetTokenGenerator,
+                                url_name="password_reset_confirm") -> str:
     token = token_generator.make_token(user_profile)
     uid = urlsafe_base64_encode(force_bytes(user_profile.id))
-    endpoint = reverse('password_reset_confirm',
+    endpoint = reverse(url_name,
                        kwargs=dict(uidb64=uid, token=token))
     return f"{user_profile.realm.uri}{endpoint}"
 
diff --git a/zerver/views/home.py b/zerver/views/home.py
index 916a6a3fa6..d7e77bcc21 100644
--- a/zerver/views/home.py
+++ b/zerver/views/home.py
@@ -167,8 +167,13 @@ def home_real(request: HttpRequest) -> HttpResponse:
     # with the code in `authenticate` of EmailAuthBackend. Here, we redirect them
     # to password reset form stating that the user's password is too weak.
     if request.session.get('needs_to_change_password', False):
-        return HttpResponseRedirect(generate_password_reset_url(user_profile,
-                                    default_token_generator).replace('password', 'weak_password'))
+        return HttpResponseRedirect(
+            generate_password_reset_url(
+                user_profile,
+                default_token_generator,
+                url_name="change_weak_password"
+            )
+        )
 
     narrow, narrow_stream, narrow_topic = detect_narrowed_window(request, user_profile)
 
```

mateuszmandera commented on zerver/tests/test_auth_backends.py:
```diff
@@ -544,6 +544,7 @@ def _mock_key(self: RateLimitedAuthenticationByUsername) -> str:
 
         def attempt_authentication(username: str, password: str) -> Optional[UserProfile]:
             request = HttpRequest()
+            request.session = self.client.session
```
Since the `client` has nothing to do with what we're doing in this function, I think this should be set to a newly made instance of session rather than `client.session`.

chdinesh1089 replied:
Used `Client().session` as that felt like the only better way. We cannot make `request=mock.MagicMock` as some rate limit code requires it to be a proper request. So, having a proper session felt better rather than using mock just for the session.

mateuszmandera replied:
We can use  a `MagicMock` for `request.session`, I think that'll be much less confusing than assigning `client.session` to it

chdinesh1089 replied:
Thanks for demonstrating this @mateuszmandera 
I was looking [this file](https://github.com/django/django/blob/master/django/test/testcases.py#L155) and ended up concluding that's the better way. I should've had a good look into `django/tests/client.py`.

(changed to use `MagicMock`)

mateuszmandera commented on zproject/backends.py:
```diff
@@ -351,10 +352,15 @@ def authenticate(self, request: HttpRequest, *,
-            if hasattr(request, 'session'):
-                request.session['needs_to_change_password'] = not check_password_strength(password)
-                return user_profile
+        try:
```
this can be made nicer by starting with `is_password_correct = user_profile.check_password(password)` - then you only need to have that in the

chdinesh1089 replied:
That looks great. Changed it. Thanks!

mateuszmandera commented on zproject/backends.py:
```diff
@@ -351,10 +352,15 @@ def authenticate(self, request: HttpRequest, *,
-                return user_profile
+        try:
+            if user_profile.check_password(password):
+                if hasattr(request, 'session'):
```
hmm I might be confused at this point, but didn't we end up deciding in a chat to get rid of the `hasattr` check and just assume `request` has a session?

chdinesh1089 replied:
Oops, sorry. I thought I haven't committed that change. Changed it.

mateuszmandera commented:
" This function is never called without None" in the 2nd commit message, I believe you meant sth like " This
function is never called without a request"

mateuszmandera commented on zproject/backends.py:
```diff
@@ -352,6 +352,7 @@ def authenticate(self, request: HttpRequest, *,
         if user_profile is None:
             return None
         if user_profile.check_password(password):
+            request.session['needs_to_change_password'] = not check_password_strength(password)
```
Hmm, let's not set a session variable unless necessary - so only if the password is too weak, because it will be very rare - and doing this might generate an additional database, or at least cache, write.

chdinesh1089 replied:
makes sense to me. Changed.

timabbott replied:
We need to clear the session variable if it's set, though, right?

chdinesh1089 commented:
> " This function is never called without None" in the 2nd commit message, I believe you meant sth like " This
> function is never called without a request"

Ohh, thanks! I meant to write request. Changed.
- - -

@timabbott This is now ready for your review.

chdinesh1089 commented:
Resolved conflicts.

timabbott commented:
I merged a couple prep commits and pushed back after fixing the `.replace` thing noted above.  I think we still need to address the following:
* How does `needs_to_change_password` get cleared on the session?  I think we want to clear it if it were previously set.
* For the last commit, why does `PasswordTooWeakError` behave differently from `if not check_password_strength`?  It seems like they should have the same outcome, or am I missing something?

chdinesh1089 commented:
Thanks for the review @timabbott 
> * How does `needs_to_change_password` get cleared on the session? I think we need to delete it from the session if it were previously set when a new password is set.

Didn't think of this as the manual testing went well, but on investigating a bit on how the session was getting cleared, found that the session is being reset due to the change of password. (by calling `user_profile.set_password(password)` which logs the user out which in turn changes session hash. this has some info: https://docs.djangoproject.com/en/3.1/topics/auth/default/#session-invalidation-on-password-change )



> * For the last commit, why does `PasswordTooWeakError` behave differently from `if not check_password_strength`?  It seems like they should have the same outcome, or am I missing something?

I should've mentioned the reason behind this. The reasoning is:
> `PasswordTooWeakError` is not an exception you can trust to mean "sure, they user gave a good password". It works right now, but that's relying on an implementation detail and there's nothing about that exception to keep it working this way. We can't trust that to log the user in

(Thanks to Mateusz Mandera for that. I haven't thought about it becoming a security threat but it really can be.)

Can we look into the last one later? (I haven't been getting much time from a few days with college going on)

timabbott commented:
Closing in favor of the rebased #18389.  Thanks @chdinesh1089 for writing this!

mateuszmandera commented:
@timabbott We missed one commit from this PR (actually the main one) - it wasn't pushed in #18389 because it's tangential to the `PasswordTooWeakError` bug, but its actually the essence of this PR - so I'm reopening for now, with just the rebased commit.

timabbott commented on templates/zerver/reset_confirm.html:
```diff
@@ -16,6 +16,9 @@ <h1>{{ _('Set a new password.') }}</h1>
             <form method="post" id="password_reset" class="form-horizontal" autocomplete="off">
                 {{ csrf_input }}
+                {% if weak_password %}
+                <p class="weak_password">Your current password is too weak for this organization's password security policy.</p>
```
This should be tagged for translation.

timabbott commented on zproject/urls.py:
```diff
@@ -547,6 +547,16 @@
+            extra_context={"weak_password": True},
+        ),
+        name="change_weak_password",
+    ),
```
Why do we need a separate URL, rather than just an extra parameter `?weak_password=True`, for example?

chdinesh1089 replied:
We cannot use GET parameters for `PasswordResetConfirmView` because of how it was written.

This is how it works along with the what's restricting us from using `?weak_password=True`:
* When the browser requests `accounts/weak_password/reset/<uidb64>/<token>/`, django saves the `token` in URL to session and then redirects to `accounts/weak_password/reset/<uidb64>/set-password`.*(this is where our GET params get lost as django uses `request.path` to get our orignal URL(`accounts/weak_password/reset/`), but `request.path` doesn't include GET params)*
* Now, our "set a new password" template is rendered. (This is where we need that `weak_password: True` to tell the user their password is weak). On entering a new password and sending a post request, django validates the `token` from session and then resets the user's password.

chdinesh1089 commented on static/styles/portico/portico.css:
```diff
@@ -1192,6 +1192,10 @@ input.new-organization-button {
 
+    .weak_password {
+        margin-bottom: 20px;
+    }
```
sure:
![image](https://user-images.githubusercontent.com/44665669/117696284-c0c62800-b1de-11eb-973a-6b4d52abc4a3.png)
(we also had some discussion regarding this [here](https://chat.zulip.org/#narrow/stream/3-backend/topic/Checking.20password.20strength.20on.20login/near/911399) :)

chdinesh1089 commented on zproject/backends.py:
```diff
@@ -422,6 +422,8 @@ def authenticate(
 
         if is_password_correct:
+            if not check_password_strength(password):
+                request.session["needs_to_change_password"] = True
```
We wanted to raise a JSON error saying 'You need to reset your password.' for those cases. (Seeing commit message, we probably had code for that in earlier versions but was removed accidentally)

Made that change and added a test at `test_auth_backends.AuthBackendTest.test_email_auth_backend_weak_password_no_session`

timabbott commented on zerver/tests/test_auth_backends.py:
```diff
@@ -342,6 +342,23 @@ def test_email_auth_backend_disabled_password_auth(self) -> None:
+                    realm=get_realm("zulip"),
+                )
+            except JsonableError as e:
+                self.assertEqual(str(e), "You need to reset your password.")
```
Let's replace this with an end-to-end test of actually attempting mobile authentication, not just calling the auth backend.

chdinesh1089 replied:
Added two tests similar to the existing `AuthBackendTest.test_email_auth_backend_password_hasher_change` and `FetchAPIKeyTest.test_old_weak_password_after_hasher_change`.

Not quite sure how to write a test for email auth with mobile. Do we have anything similar doing email auth for mobile for me to use as reference or any other pointers that'd help? Wasn't able to find any tests for mobile flow other than the ones in social auth tests.

timabbott commented on zproject/backends.py:
```diff
@@ -422,6 +422,12 @@ def authenticate(
+                if hasattr(request, "session"):
+                    request.session["needs_to_change_password"] = True
+                else:
+                    raise JsonableError(_("You need to reset your password."))
```
I kinda feel like it'd be better for this to share its code path with `PasswordTooWeakError`; it's not worth having two different mechanisms for what happens when you're trying to login and your password is too weak.

I think we can just do
```
if is_password_correct and not check_password_strength(password):
     raise PasswordTooWeakError()
```

in the above block, and remove most of the other new code in this pull request.

The UX will be a little less friendly, in requiring a reset rather than letting the user change their password without the reset link flow, but it'll also be more secure, and avoid extra duplicated code paths here.

chdinesh1089 replied:
Okay, but that will leave us unable to figure out if the `PasswordTooWeakError` was from the rehashing issue for us to be able to log that. Ended up using nested try, except blocks to be able to log that corner case.

chdinesh1089 committed:
auth: Force users to change weak passwords.

Doesn't let the user login until they change their password
similar to what is done when PasswordTooWeakError is raised due
to hasher change.
